akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -346,9 +349,12 @@ final class Split[T](decision: Split.SplitDecision, p: T ⇒ Boolean) extends Gr
             closeThis(handler, elem)
             handOver(handler)
           } else {
-            // Drain into the void
-            if (substreamCancelled) pull(in)
-            else substreamSource.push(elem)
+            if (substreamCancelled) {
+              if (eagerClose) completeStage()
```

Comment:
Hmm, not sure I completely understand: receiving cancellation from downstream should only happen if the downstream decides that it does not want more data, which means that canceling is the desired semantics in the eagerClose case (which BTW should be renamed to eagerCancel). Normal substream termination due to the next element matching the substream will signal `substreamSource.complete()`, inhibiting further reception of `onDownstreamFinish()`, so in that case all should be good.

The only “interesting” case is if the substream processing logic double-checks the length of the stream and terminates of its own accord just at the point where the stream would otherwise have ended—but I consider that a less important case than to pass along cancellation without needing further upstream data; an open TCP connection binds resources even when no data are flowing.